### ISSUE TITLE
Cycle 3 Funding: Develop a Jupyter widget-based app for reducing and performing stellar photomtery on optical images

### ISSUE THREAD

**Issue by jdswinbank** [astropy/astropy-project#311]
This issue captures execution of the Cycle 3 project “Devvelop a Jupyter widget-based app for reducing and performing stellar photomtery on optical images” (#279). The full text of the proposal is available [here](https://github.com/astropy/astropy-project/blob/main/finance/proposal-calls/cycle3/optical-reduction-photometry-entry-level-gui-app.md).

This project was funded to the level of $32,000.

**Comment by jdswinbank** [comment]
Hey @mwcraig — now that the funding for this project has been approved, it would be a good idea to start drawing up a bit more concrete plan for how you'd like to disburse the funds. I understand from the proposal that you will lead, so we probably want to start with deciding how much of the budget you are going to need personally and then I can start the ball rolling with a scope of work etc.

**Comment by mwcraig** [comment]
Would it be ok to start working on this next week (after Nov 7) -- I'm heading to a conference in a few days.

**Comment by kelle** [comment]
I think I fixed it. Photometry , right?

**Comment by pllim** [comment]
Yes, typo is fixed. Thanks!

**Comment by jdswinbank** [comment]
I hate to be picky, but the [approved proposal](https://github.com/astropy/astropy-project/blame/main/finance/proposal-calls/cycle3/optical-reduction-photometry-entry-level-gui-app.md#L3) is for an app to perform photomtery. If we get some photometry tool out of it, I'm not sure it really does the job...

**Comment by jdswinbank** [comment]
> Would it be ok to start working on this next week (after Nov 7) -- I'm heading to a conference in a few days.

No problem. Please post on this issue whenever you're ready.

**Comment by jdswinbank** [comment]
Hey @mwcraig — just a ping. No need to rush, but it would be nice to get started on this when you're ready.

**Comment by mwcraig** [comment]
@jdswinbank -- I wasn't sure whether you wanted what I had in mind here as a comment or not, so I put it into a gist for now: https://gist.github.com/mwcraig/4f42b512997333e79948dd5cae5e50b8

**Comment by jdswinbank** [comment]
Hi @mwcraig — thanks for that; a very nice summary.

I had previously assumed that we'd press ahead with a contract/subaward for you personally, since the original proposal lists you as leading this project. However, I understand from the disbursement proposal that the first step is actually to hire a third party to start work. That's not a problem; I'll put thoughts of a contract for you on hold for now.

After we have all the contracts for existing project members sorted out (and we're nearly done now) the Finance Committee will put our heads together and come up with some guidance for hiring third parties. We'll get back to you on this shortly (I'm guessing that'll be early next year).

**Comment by mwcraig** [comment]
This is an updated request in response to the call for additional funding opportunities. As reminder, the amount of the original request was $40,000, and the amount approved was $32,000.

**The updated request is for $58,025, as detailed below.** Part of the money would be used to hire me or people I've identified directly as independent contractors, and part would be a subaward for MSUM (my institution) to be spent Fall 2023.

Since this is such a substantial increase I have specified a priority for funding for each of the items below.

## Area 1: Code development and guide writing
### Priority: High
### Amount: $36,000 (360 develop hours at $100/hour)
### Description

As explained in [this gist](https://gist.github.com/mwcraig/4f42b512997333e79948dd5cae5e50b8), the two main goals of this project are to develop a photometry guide and to develop a general-purpose photometry tool for stellar astronomy, particularly aimed at relatively novice users.  

The initial phase of this work is to restructure the package [`stellarphot`](https://github.com/feder-observatory/stellarphot), which I've used for the last couple of years in an introductory observational astronomy course, to have a simpler and better documented and tested API, In parallel with that, work will begin on the photometry guide.

I am proposing that we hire, as independent contractors over the summer, **at the rate of $100/hour**:

| Person | Hours | Justificaton |
| :------: | :----: | :-----------: |
| Matt Craig | 160 | Write the photomtry guide, oversee the restructuring of `stellarphot`, and help with that restructuring. |
| Contractor[1] | 160 |  Propose new structure for  `stellarphot` and assist in much of the code re-write. |
| Contractor[2] | 40 | Act as beta tester for revamped `stellarphot` and photometry guide. |

[1] There is a colleague of mine with a background in computer science and astronomy that I have identified as an excellent fit for this role.

[2] Another colleague with background in astronomy and some familiarity with Python has been identified as an excellent fit for this role.

*In addition*, undergraduate students paid with separate funds will participate in testing over the summer and are learning how to make code contributions.

## Area 2: Compensation for undergraduates at MSUM in Fall 2023 for ongoing testing and bug fixes
### Priority: Medium
### Amount[3]: $10,125 ($15/hour for 600 hours, plus 12.5% overhead and 7.65% for fringe benefits)

### Description

The photometry guide and `stellarphot` will be used in an observational astronomy course this fall. I anticipate usage will uncover bugs in `stellarphot` and areas of the guide that are unclear or incorrect.

The students hired with these funds will fix those bugs and help with updating the guide. 

[3]: The overhead and fringe rates are approximate and will need to be confirmed before the subaward is made. The total budget for this area will not exceed $10,125 and the pay rate will not be less than $15/hour. The number of hours will be slightly adjusted, if necessary.

## Area 3: Teaching release for Matt Craig 
### Priority: Low
### Amount[4]: $11,900 (2 credits of teaching release at $3872/credit, with 12.5% overhead and 40% for fringe benefits, and 1% pay increase)
### Description

Though part of my teaching load in Fall 2023 will be observational astronomy, I will also be teaching several other courses. Two credits of teaching release will provide an additional 6 hours per week to work on fixing bugs and updating the photometry guide as the semester progresses.

[4]: The overhead and fringe rates are approximate and will need to be confirmed before the subaward is made. In addition, there will likely be a pay increase when our new contract is negotiated this summer. The total budget for this area will not exceed $11,900. The number of credits of teaching release will be adjusted, or the money will be used to pay for additional duty days during the Fall semester.

**Comment by kelle** [comment]
To be clear, the request is for an additional $26,025, correct. Bringing the total budget from $32k to $58k?

**Comment by mwcraig** [comment]
Yes, though the highest priority is an increase of $4k to $36k for the summer pay. 

Some money for undergrads would be helpful. 

Whether I get reassigned time is really a very low priority.

**Comment by jdswinbank** [comment]
Hey @mwcraig — the Finance Committee has approved an increase to the allocated funding of $26,025 for this project for a total of $58,025. Please submit invoices to the Moore grant as described above. We encourage you to invoice for all this work, including the extension, by the end of December if possible, and certainly not later than 31 March 2024.

**Comment by mwcraig** [comment]
@jdswinbank -- wonderful, thanks for the news!

1. I am working with my campus to get our internal paperwork ready for Areas 2 and 3 of the revised proposal. What additional steps will need to be taken to get that subaward set up?
2. The money for Area 1 is meant to be paid directly to the folks doing the work. I can ask them to set up accounts in OpenCollective. What additional steps beyond that need to be taken before submitting invoices?

**Comment by jdswinbank** [comment]
As I understand it, the plan is that:

- Each of the individuals working on Area 1 will receive a separate contract with NumFOCUS.
- MSUM will receive a subaward from NumFOCUS covering all the work to be performed in Areas 2 & 3.

The process for both the contracts and the subaward looks similar from our point of view.

1. We — you and I — will agree on brief (couple of paragraph) statements of work describing what the award/contract will cover. The text you've provided above should pretty much cover this; I will do some editing and propose some text for you to  approve. We should also agree start/end dates for the contract.
2. We will pass these to NumFOCUS and ask them to begin the contracting process.
3. NumFOCUS will take it from there, reaching out to the funded entities with a contract for their signature.
4. Everybody will need to be registered not just with OpenCollective, but also with Rocket (https://numfocus.typeform.com/to/umyusM). This is pretty straightforward, and we/NumFOCUS can provide help as necessary.

I hope that covers everything — let me know if you have questions, and please ping me if you've not received draft SOWs within the next few days.

**Comment by mwcraig** [comment]
Just a note that I've also asked MSUM about administering Area 1, in the event they are able to do that with minimal deductions to minimize the number of contracts.

**Comment by jdswinbank** [comment]
Here's a draft of the SoW. I've written this on the assumption that all of this work takes place through an award to MSUM; we can split off work package 1 (your area 1, above) into separate contracts if necessary.

[Craig _ MSUM SOW.pdf](https://github.com/astropy/astropy-project/files/12213166/Craig._.MSUM.SOW.pdf)

I think this is all pretty straightforward — it's basically just reiterating what you've said above — but please yell if anything looks incorrect.

**Comment by mwcraig** [comment]
@jdswinbank -- this looks good.

A couple of updates from our end: 

- The fall money that was to be used for teaching release for me will instead be used to augment my pay in the fall (it was too late to hire a replacement). My load includes an observational astronomy course in which we will use the material developed under the contract.
- Once we moved all of the summer salary into the university award, the costs increased somewhat. I am negotiating with the university to reduce those costs, but said i would ask about a couple of possibilities. "No" is a perfectly fine answer to both of these.
    1. Would Astropy consider increasing the grant by an additional $8,000, to $66,000?
    2. Would Astropy consider increasing the grant by an additional $1,000, to approximately $59,000?

**Comment by kelle** [comment]
What is the motivation for the increases? I'm guessing maybe fringe or overhead rate change?

**Comment by mwcraig** [comment]
Sort of; by running the pay through the university more of it is subject to overhead.

**Comment by kelle** [comment]
Can you provide us with a break down of the budget indicating direct and indirect/overhead costs?

**Comment by mwcraig** [comment]
@kelle and the rest of the Finance Committee -- some good news. MSUM has decided to use its overhead to partially defray the cost of the fringe benefits in the restructured grant.

As result, the amount of the increase I am requesting is $698.17, bringing the total to $58,723.

A detailed accounting of the cost is [in this google sheet](https://docs.google.com/spreadsheets/d/1rHtccGpQXdITsgG1SCeNw_M_-_tvWBwn1e3AfXdhgQ4/edit?usp=sharing).

The work hours dedicated to the grant activities are still roughly 6.5% higher than in the original proposal, as explained below. With my administration's commitment to dedicate the overhead to paying fringe benefits to cost to Astropy for this increase in hours would be the amount of the increase I am requesting, $698.17.

> Including fringe benefits in the Astropy funding means that the folks who have been doing the summer work do not have to pay self-employment tax. They have agreed to work more hours at a reduced rate to recognize that increase in their take home pay. The increase in hours is about 6.5%.

**Comment by mwcraig** [comment]
I'm giving this a friendly bump -- I'm wondering if a decision has been made about the additional increase.

**Comment by jdswinbank** [comment]
Hi Matt — sorry for the delay on this one. We approve this increase of $698.17, for a total budget of $58,723, and apolgise for how long it took us to get here!

**Comment by mwcraig** [comment]
@jdswinbank -- we should have invoices for some of this working on their way soon. I wanted to ask one question: What is the last day that work for this can be invoiced?

**Comment by jdswinbank** [comment]
The official end date of the Moore grant, which is funding this work, is 2024-04-01, so that's an absolute hard deadline. In order to avoid a last minute panic, though, if possible I'd encourage you to submit well in advance of that. We have to report to Moore on 1 March, and we'd love to have as much as possible finalized by then.

**Comment by mwcraig** [comment]
I wanted to provide a brief update on the status of this work prior to submitting a final report at the end of March.

- The package `stellarphot` is close to an alpha release of version 2.0. We are testing internally and I anticipate the alpha will be out early the week of Mar 11. A few highlights:
    -  Once settings have for your camera, observatory, etc., doing photometry on an image ro folder of images is a couple of lines of code.
    - Rewriting the code to work with pydantic 2, which was released in early Fall 2023 took much more effort than anticipated. It turns out to be much more flexible for adding adding pydantic behavior to non-pydantic classes like astropy's `Time` or `SkyCoord` or `Quantity`. The upside is that we've worked out how to use pydantic to serialize those objects, generate json files of settings for use in stellarphot, and automatically generate widget-based forms for entering values.
    - The AAVSO has expressed interest in using stellarphot to perform photometry on images gather by amateurs using highly automated telescopes like those made by Unistellar and ZWO's SeeStar. 
    - I have names/emails of roughly 20 AAVSO members from the Astropy workshop presented at the AAVSO meeting last November who are willing to try out the code. That will start this month and continue well beyond that, I think.

- The photometry guide is (finally) coming along nicely. Look for a large-ish PR the week of Mar 11, with updates to an open PR prior to that.
    -  Though the jupyter-book project on which the guide builds has recently released the first stable version of the package, I plan to wait to update the guide to that version until early April so as to ensure content gets in before then.

**Comment by mwcraig** [comment]
I am overdue for an update. 

+ The photometry guide is still coming along. I made a new branch in the repository if you would like to look at the current state. I ened up updating the jupyter-book version, in part to add a feature that comments may be made on any of the pages. Those comments are turned into github issues. Though this isn't ready for broad distribution and feedback yet, you can look at the photomery sections here: https://www.astropy.org/ccd-reduction-and-photometry-guide/v/pdev

+ The list of issues remaining in to be resolved in stellarphot is down to a handful, listed [here](https://github.com/feder-observatory/stellarphot/issues?q=is%3Aopen+is%3Aissue+milestone%3A%222.0.0+--+Reimplemented+data+structure%22). Almost all of them are re-writes of the notebook interface, which are straightforward now that the underlying infrastructure is complete.

**Comment by kelle** [comment]
These funds have been spent.
